rfc/20240725-exclude-resources.md

```diff
@@ -0,0 +1,121 @@
+# Exclude Flag for Planning and Applying
+
+Issue: https://github.com/opentofu/opentofu/issues/426
+
+The RFC entails a new flag `-exclude` to be used in planning and applying. The flag's purpose is to be the inverse of the `-target` flag - A targeted plan/apply for every resource that is not excluded
+
+This is solving many problems, like some multi-stage deployment configurations, as well as edge cases where you would just like to skip applying a specific resource temporarily
+
+## Proposed Solution
+
+An `-exclude` flag that could be used in `tofu plan`, `tofu apply` and `tofu destroy`.
+
+This flag will work as an exact inverse of `-target` - when planning, it would act as though we are targeting any resource that is not excluded.
+
+Similarly to how targeted resources also include all of their dependencies in the plan, an excluded resource would mean that all resources dependent on it should be excluded as well
+
+### User Documentation
+
+User should be able to provide one or more excluded resource, via one or multiple `-exclude` flags. For example: `tofu plan -exclude=null_resource.a -exclude=null_resource.b`
+
+An excluded plan - Would exclude any resource given via the `-exclude` flag, and also any resource that is dependent on these resources.
+
+See the following example:
+
+```hcl
+# In this example there are 4 null_resources, with the following dependency graph:
+# A <---- [B,C] <---- D
+
+resource "null_resource" "a" {
+  
+}
+
+resource "null_resource" "b" {
+  triggers = {
+    a = null_resource.a.id
+  }
+}
+
+resource "null_resource" "c" {
+  triggers = {
+    a = null_resource.a.id
+  }
+}
+
+resource "null_resource" "d" {
+  triggers = {
+    b = null_resource.b.id
+    c = null_resource.c.id
+  }
+}
+```
+
+With the above example:
+- Running `tofu plan -exclude=null_resource.d` would plan any resource that's not `null_resource.d` (`null_resource.a`, `null_resource.b`, `null_resource.c`)
+- Running `tofu plan -exclude=null_resource.a` would create an empty plan, since all resources depend on `null_resource.a`
+- Running `tofu plan -exclude=null_resource.b` would exclude both `null_resource.b` and `null_resource.d` which depends on it (so it will plan `null_resource.a` and `null_resource.c`)
+- Running `tofu plan -exclude=null_resource.b -exclude=null_resource.c` would exclude `null_resource.b`, `null_resource.c` and also `null_resource.d` which depends on one of them (or in this case - both of them)
+- Running `tofu plan -exclude=null_resource.a -exclude=null_resource.b` would create an empty plan, since all resources depend on `null_resource.a`
+
```

Comment:
Not a terribly important aspect, but I'm wondering how `-exclude=null_resource.e` would behave, since it's a non-existent resource. Would it get ignored, or would the command error out?

Reply:
Personally, I'd just have it be ignored, and plan will be full. That would provider parity with how `-target` works, which `-exclude=null_resource.e` would result in an empty plan.

I'll update info on this in the RFC

Reply:
Updated aa91bf5c45e109f595e07e64b493323ef0b6a802